### DISCUSSION TITLE
Optimizing Token Generation in llama.cpp's CUDA Backend

### DISCUSSION THREAD

**Discussion by am17an** [ggml-org/llama.cpp#17621] [Show and tell]
LLM inference is divided into 2 phases: prompt-processing (PP) and token-generation (TG). Prompt processing is when entire prompt is fed through the model, whereas token-generation is when the model starts outputting one token at a time. These are different workloads, for decoder-only auto-regressive LLMs PP is compute-bound and TG is memory-bound. 

This post covers optimizations for TG using kernel fusion and concurrent streams in the CUDA backend for llama.cpp. Other backends also do fusion but differently, and the rest of this post assumes the CUDA backend. This post is lightly technical with links to PRs for people who want to go into further detail.

I ran tests on 4090 and 5090 with baseline as these optimizations disabled.

`GGML_CUDA_GRAPH_OPT=1 ./llama-bench -m qwen3_30b_qk_m.gguf,gpt_oss_20b-mxfp4.gguf -fa 1 -p 0`

| model                          |   device |      size | params  | backend | ngl |  fa |  test | t/s           |
| ------------------------------ | -------: | --------: | ------- | ------: | --: | --: | ----: | ------------- |
| qwen3moe 30B.A3B Q4_K - Medium | RTX 5090 | 17.28 GiB | 30.53 B |    CUDA |  99 |   1 | tg128 | 352.06 ± 1.76 |
| gpt-oss 20B MXFP4 MoE          | RTX 5090 | 11.27 GiB | 20.91 B |    CUDA |  99 |   1 | tg128 | 419.14 ± 0.89 |
| qwen3moe 30B.A3B Q4_K - Medium | RTX 4090 | 17.28 GiB | 30.53 B |    CUDA |  99 |   1 | tg128 | 271.04 ± 0.65 |
| gpt-oss 20B MXFP4 MoE          | RTX 4090 | 11.27 GiB | 20.91 B |    CUDA |  99 |   1 | tg128 | 271.99 ± 0.53 |

vs baseline`GGML_CUDA_DISABLE_FUSION=1 GGML_CUDA_GRAPH_OPT=0 ./llama-bench -m qwen3_30b_qk_m.gguf,gpt_oss_20b-mxfp4.gguf  -fa 1 -p 0`

| model                          | device    |     size |  params | backend | ngl |  fa |  test |           t/s |
| ------------------------------ | --------: | ------: | ------- | --: | --: | ----: | ----:| ------------: |
| qwen3moe 30B.A3B Q4_K - Medium | RTX 5090  | 17.28 GiB | 30.53 B | CUDA    |  99 |   1 | tg128 | 246.96 ± 0.31 |
| gpt-oss 20B MXFP4 MoE          | RTX 5090  | 11.27 GiB | 20.91 B | CUDA    |  99 |   1 | tg128 | 329.23 ± 0.16 |
| qwen3moe 30B.A3B Q4_K - Medium | RTX 4090  | 17.28 GiB |    30.53 B | CUDA       |  99 |  1 |           tg128 |        198.39 ± 0.05 |
| gpt-oss 20B MXFP4 MoE          | RTX 4090  | 11.27 GiB |    20.91 B | CUDA       |  99 |  1 |           tg128 |        232.05 ± 0.02 |

According to my rough napkin math [^1], the theoretical speed-of-light for `gpt-oss-20b` on a 5090 is ~520 t/s, and we are able to reach ~419 t/s (~80%), although speedups are more modest as context size grows.

Most of the optimizations are enabled by default in llama.cpp, concurrent streams is enabled via an environment flag `GGML_CUDA_GRAPH_OPT=1` (note that this particular optimization only works for a single GPU at the moment, also note that this has nothing to with CUDA Graphs)

## Kernel Fusion

Fusing kernels reduces memory traffic and kernel launch time. Since TG is memory-bound, kernel fusion helps there the most. From experience fusion is the most useful when we have high memory bandwidth, conversely the results are not as great when the bandwidth is low.  

Note that some fusion will increase register pressure and generally leads to more complex kernels, so it's not always worth doing.

llama.cpp's backend is based on ggml, a low level tensor library. ggml offers a lot of flexibility to backends via its cgraph (compute graph) interface. After upfront tensor allocation and initial graph ordering, it passes control to individual backends to compute the graph. The graph itself is composed of `ggml_tensor`s, which contain all the information required for fusion.

The backends are then responsible for fusion. The CUDA backend inspects order of operations and decides whether to fuse a bunch of individual kernels. This pattern can be described in pseudo-code:

```
for (int i = 0; i < graph->n_nodes; ++i) {
	if (can_fuse(graph, i, FUSION_PATTERN)) { 
		replace_nodes_with_kernel(graph, i, FUSION); 
		i += SKIP_NODES; // Skip the nodes we just absorbed 
	}
}
```

In ggml-cuda, the relevant portions start from [here](https://github.com/ggml-org/llama.cpp/blob/fa0465954faef9d7170b967ad89f8bc5303a32f3/ggml/src/ggml-cuda/ggml-cuda.cu#L3294)

Looking at the order of nodes, one can identify opportunities for fusion. Something which is commonly used in a lot of models is usually a good candidate. The following are the most effective fusions that the CUDA backend does at the moment:

### GEMV fusion 

Matrix-vector multiplication is used instead of matrix-matrix multiplication during TG as we operate on 1 token at a time. We identify operations where `MUL_MAT` follows an `ADD` or more interestingly, when there is a gated activation i.e. the operation  $o = \sigma(W_{gate} X)\odot W_{up}X$, so we can re-use the X activation and multiply by both matrices. More details [here](https://github.com/ggml-org/llama.cpp/pull/16715) 

### TopK-MoE

This is the common scoring algorithm in MoE models to select expert weights per token. This usually involves a soft-max over the logits per expert, then a top-k selection for the experts for the sparse MoE matrix multiplication (called MUL_MAT_ID in ggml). More details [here](https://github.com/ggml-org/llama.cpp/pull/16130)

### Fused RMS Norm

This is a relatively simple one, usually `RMS_NORM` follows a `MUL` and optionally an `ADD`. We just do everything in the same kernel. More details [here](https://github.com/ggml-org/llama.cpp/pull/14800). 

Others include fusing [adds](https://github.com/ggml-org/llama.cpp/pull/15631), fusing the [softcap](https://github.com/ggml-org/llama.cpp/pull/14907) operation.

## Concurrent CUDA Streams

> Note: At the moment to run this you need to supply the env variable `GGML_CUDA_GRAPH_OPT=1`. 

Since query (Q), key (K), and value(V) projections in the attention block are independent of each other prior to calculating the attention scores, we can run them independently using CUDA streams. This is better explained with a NSight diagram.
<img width="1024" height="480" alt="qkv concurrent streams" src="https://github.com/user-attachments/assets/c460d4c4-d105-40f5-9aff-090036c9f204" />

The ggml graph allocator optimizes for memory footprint by reusing 'expired' buffers immediately. In a multi-stream context, 'expired' is ambiguous, leading to race conditions where Stream B overwrites Stream A's input. This is currently "solved" via interleaving the graph nodes to extend their lifetime. There will be a better long-term fix in ggml to allow for such scenarios. More details [here](https://github.com/ggml-org/llama.cpp/pull/16991)

## Conclusion 

Although none of these PRs increase the TG by more than 10%, taking them all altogether we get a nice speedup.

## Acknowledgements

[Johannes](https://github.com/JohannesGaessler), [Jeff](https://github.com/jeffbolznv), [Georgi](https://github.com/ggerganov), [Oliver](https://github.com/ORippler), [Sigbjørn](https://github.com/CISC)

[^1]:  speed-of-light (SoL) [napkin math](https://gist.github.com/am17an/6833d30230b877c93abf33ca12a07937)

**Comment by CISC** [comment]
The NSight diagram link doesn't work (also took the liberty to edit my name since everyone else was on a first-name basis). :)

**Reply by am17an** [reply]
Fixed, and I always thought your first name was CISC! (just kidding...)

**Comment by coder543** [comment]
I think your napkin math for the theoretical limit could be a little pessimistic.

```
$ llama-bench -m ggml-gpt-oss-20b-mxfp4.gguf -ngl 999 -fa 1 -p 2048 -n 32 -mmp 0 -r 5
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GB10, compute capability 12.1, VMM: yes
| model                          |       size |     params | backend    | ngl | fa | mmap |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | -: | ---: | --------------: | -------------------: |
| gpt-oss 20B MXFP4 MoE          |  11.27 GiB |    20.91 B | CUDA       | 999 |  1 |    0 |          pp2048 |      4317.51 ± 29.87 |
| gpt-oss 20B MXFP4 MoE          |  11.27 GiB |    20.91 B | CUDA       | 999 |  1 |    0 |            tg32 |         86.06 ± 0.64 |
```

With 273GB/s / 3.45GB, this would only yield 79 tokens per second _theoretical_, but I'm seeing 86 in practice using the `ggml-org` quant of `gpt-oss-20b` on the Nvidia DGX Spark.

I had `codex` consider the problem, and it came up with a script: [gguf_bandwidth.py](https://github.com/user-attachments/files/24402743/gguf_bandwidth.py)

If I run this against gpt-oss-20b, I get:

```
$ gguf_bandwidth --bw 273 --kv-used 0 ggml-gpt-oss-20b-mxfp4.gguf
model: ggml-gpt-oss-20b-mxfp4.gguf
layers: 24
context_length: 131072
sliding_window: 128
head_count_kv: 8
key_len: 64
value_len: 64
kv_bytes_per_elem: 2
kv_used: 0
iswa: True
n_swa: 128
swa_layers: 12
base_layers: 12
swa_full: False
ubatch: 512
n_seq_max: 1
n_kv_base: 0
n_kv_swa: 0
moe: True
expert_count: 32
expert_used_count: 4
weights_used_bytes: 3190031616
weights_total_bytes: 12096558336
kv_read_bytes: 0
kv_write_bytes: 49152
kv_cache_bytes: 3240099840
total_mem_bytes: 15336658176
total_mem_GB: 15.336658
total_mem_GiB: 14.283376
total_bytes: 3190080768
total_GB: 3.190081
total_GiB: 2.970994
bw_GBps: 273
tps_max_theoretical: 85.5778
```

Which is not far off of the reality on the DGX Spark.

Setting --bw to 1792 for the RTX 5090:

```
$ gguf_bandwidth --bw 1792 --kv-used 0 ggml-gpt-oss-20b-mxfp4.gguf
model: ggml-gpt-oss-20b-mxfp4.gguf
layers: 24
context_length: 131072
sliding_window: 128
head_count_kv: 8
key_len: 64
value_len: 64
kv_bytes_per_elem: 2
kv_used: 0
iswa: True
n_swa: 128
swa_layers: 12
base_layers: 12
swa_full: False
ubatch: 512
n_seq_max: 1
n_kv_base: 0
n_kv_swa: 0
moe: True
expert_count: 32
expert_used_count: 4
weights_used_bytes: 3190031616
weights_total_bytes: 12096558336
kv_read_bytes: 0
kv_write_bytes: 49152
kv_cache_bytes: 3240099840
total_mem_bytes: 15336658176
total_mem_GB: 15.336658
total_mem_GiB: 14.283376
total_bytes: 3190080768
total_GB: 3.190081
total_GiB: 2.970994
bw_GBps: 1792
tps_max_theoretical: 561.7413
```

It believes that the RTX 5090 could go as high as 560 tokens per second in a purely memory-bandwidth constrained scenario.

**Reply by am17an** [reply]
I would consider an error of +/-10% to be fine when estimating. However your perf test is for tg32, whereas the numbers posted above are for tg128

**Reply by coder543** [reply]
tg32 vs tg128 isn't really meaningful if either of the numbers are above the theoretical limit, which was calculated without regard for the number of output tokens?

But, yes:

```
$ llama-bench -m ggml-gpt-oss-20b-mxfp4.gguf -ngl 999 -fa 1 -p 2048 -n 128 -mmp 0 -r 5
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GB10, compute capability 12.1, VMM: yes
| model                          |       size |     params | backend    | ngl | fa | mmap |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | -: | ---: | --------------: | -------------------: |
| gpt-oss 20B MXFP4 MoE          |  11.27 GiB |    20.91 B | CUDA       | 999 |  1 |    0 |          pp2048 |      4251.36 ± 27.66 |
| gpt-oss 20B MXFP4 MoE          |  11.27 GiB |    20.91 B | CUDA       | 999 |  1 |    0 |           tg128 |         82.64 ± 0.14 |
```

which is still above 79 tokens per second.

I mainly just thought the script was cool, since it can be used with any `gguf` file and based on my own testing appears to do a good job of accounting for a number of nuances in calculating the amount of memory being transferred for each token generated.

I did say the calculation "could be pessimistic", not that it was terribly wrong. I like to hope that this can all be optimized even more to bring the performance up even higher. A higher theoretical ceiling provides more hope that it's not impossible.

**Reply by am17an** [reply]
Right, makes sense. BTW you can use `GGML_CUDA_GRAPH_OPT=1` for some additional perf in TG, PP performance looks quite good for DGX spark

**Reply by mediouni-m** [reply]
On an Ascent GX10 with NVIDIA-Nemotron-3-Nano-30B-A3B-Q4_K_M.gguf:

Baseline CUDA:

ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GB10, compute capability 12.1, VMM: yes
| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| nemotron_h_moe 31B.A3.5B Q4_K - Medium |  22.82 GiB |    31.58 B | CUDA       |  99 |           pp512 |       2135.98 ± 9.54 |
| nemotron_h_moe 31B.A3.5B Q4_K - Medium |  22.82 GiB |    31.58 B | CUDA       |  99 |           tg128 |         61.76 ± 0.02 |

build: a89002f07 (7761)

With `GGML_CUDA_GRAPH_OPT=1`:
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GB10, compute capability 12.1, VMM: yes
| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| nemotron_h_moe 31B.A3.5B Q4_K - Medium |  22.82 GiB |    31.58 B | CUDA       |  99 |           pp512 |       2110.23 ± 7.89 |
| nemotron_h_moe 31B.A3.5B Q4_K - Medium |  22.82 GiB |    31.58 B | CUDA       |  99 |           tg128 |         54.77 ± 1.11 |

build: a89002f07 (7761)

Vulkan backend:

ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = NVIDIA Tegra NVIDIA GB10 (NVIDIA) | uma: 1 | fp16: 1 | bf16: 1 | warp size: 32 | shared memory: 49152 | int dot: 1 | matrix cores: NV_coopmat2
| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| nemotron_h_moe 31B.A3.5B Q4_K - Medium |  22.82 GiB |    31.58 B | Vulkan     |  99 |           pp512 |      1661.48 ± 12.92 |
| nemotron_h_moe 31B.A3.5B Q4_K - Medium |  22.82 GiB |    31.58 B | Vulkan     |  99 |           tg128 |         64.50 ± 1.02 |

build: a89002f07 (7761)
